File: src/weekly_intel/processing/extractor.py
"""LLM-based content extraction using Claude API."""

import logging
import os
from dataclasses import dataclass, field
from typing import Optional

import anthropic

from weekly_intel.config import get_config

logger = logging.getLogger(__name__)


@dataclass
class ExtractionResult:
    """Result of content extraction."""

    summary: str
    key_information: list[str] = field(default_factory=list)
    themes: list[str] = field(default_factory=list)
    hot_takes: list[dict] = field(default_factory=list)  # {take: str, context: str}
    entities: dict = field(default_factory=dict)  # {people: [], companies: [], technologies: []}
    raw_response: Optional[dict] = None


EXTRACTION_TOOLS = [
    {
        "name": "extract_content",
        "description": "Extract structured information from content",
        "input_schema": {
            "type": "object",
            "properties": {
                "summary": {
                    "type": "string",
                    "description": "A concise 2-3 sentence summary of the content",
                },
                "key_information": {
                    "type": "array",
                    "items": {"type": "string"},
                    "description": "List of key facts, announcements, or data points",
                },
                "themes": {
                    "type": "array",
                    "items": {"type": "string"},
                    "description": "Main themes or topics discussed",
                },
                "hot_takes": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "take": {"type": "string", "description": "The contrarian or notable opinion"},
                            "context": {"type": "string", "description": "Brief context about why this is notable"},
                        },
                        "required": ["take", "context"],
                    },
                    "description": "Contrarian views or notable opinions",
                },
                "entities": {
                    "type": "object",
                    "properties": {
                        "people": {
                            "type": "array",
                            "items": {"type": "string"},
                            "description": "People mentioned",
                        },
                        "companies": {
                            "type": "array",
                            "items": {"type": "string"},
                            "description": "Companies or organizations mentioned",
                        },
                        "technologies": {
                            "type": "array",
                            "items": {"type": "string"},
                            "description": "Technologies, products, or concepts mentioned",
                        },
                    },
                    "description": "Named entities extracted from content",
                },
            },
            "required": ["summary", "key_information", "themes", "entities"],
        },
    }
]


class ContentExtractor:
    """Extract structured information from content using Claude."""

    def __init__(self):
        config = get_config()
        api_key = config.api_keys.anthropic or os.environ.get("ANTHROPIC_API_KEY")

        if not api_key:
            raise ValueError(
                "Anthropic API key not configured. "
                "Set ANTHROPIC_API_KEY environment variable or api_keys.anthropic in config."
            )

        self.client = anthropic.Anthropic(api_key=api_key)
        self.model = config.processing.model
        self.prompts = config.prompts.extraction

    def extract(
        self,
        content: str,
        title: Optional[str] = None,
        author: Optional[str] = None,
        source_type: Optional[str] = None,
    ) -> ExtractionResult:
        """Extract structured information from content.

        Args:
            content: The content text to analyze
            title: Optional title of the content
            author: Optional author name
            source_type: Type of source (substack, gmail, youtube)

        Returns:
            ExtractionResult with extracted information
        """
        # Build context
        context_parts = []
        if title:
            context_parts.append(f"Title: {title}")
        if author:
            context_parts.append(f"Author: {author}")
        if source_type:
            context_parts.append(f"Source type: {source_type}")

        context = "\n".join(context_parts) if context_parts else ""

        # Truncate content if too long (Claude has context limits)
        max_content_length = 50000
        if len(content) > max_content_length:
            content = content[:max_content_length] + "\n[Content truncated...]"

        user_message = self.prompts.user_prompt_template.format(
            context=context, content=content
        )

        try:
            response = self.client.messages.create(
                model=self.model,
                max_tokens=4096,
                system=self.prompts.system_prompt,
                tools=EXTRACTION_TOOLS,
                tool_choice={"type": "tool", "name": "extract_content"},
                messages=[{"role": "user", "content": user_message}],
            )

            # Parse tool use response
            for block in response.content:
                if block.type == "tool_use" and block.name == "extract_content":
                    data = block.input
                    return ExtractionResult(
                        summary=data.get("summary", ""),
                        key_information=data.get("key_information", []),
                        themes=data.get("themes", []),
                        hot_takes=data.get("hot_takes", []),
                        entities=data.get("entities", {}),
                        raw_response=data,
                    )

            # Fallback if no tool use found
            logger.warning("No tool use in response, using text content")
            text_content = ""
            for block in response.content:
                if hasattr(block, "text"):
                    text_content += block.text

            return ExtractionResult(
                summary=text_content[:500] if text_content else "Failed to extract summary",
                raw_response={"text": text_content},
            )

        except anthropic.APIError as e:
            logger.error(f"Claude API error: {e}")
            raise

    def extract_batch(
        self,
        items: list[dict],
        batch_size: int = 5,
    ) -> list[ExtractionResult]:
        """Extract from multiple items.

        Args:
            items: List of dicts with 'content', 'title', 'author', 'source_type'
            batch_size: Not used for Claude (no batching), but kept for interface

        Returns:
            List of ExtractionResults
        """
        results = []
        for item in items:
            try:
                result = self.extract(
                    content=item.get("content", ""),
                    title=item.get("title"),
                    author=item.get("author"),
                    source_type=item.get("source_type"),
                )
                results.append(result)
            except Exception as e:
                logger.error(f"Failed to extract from {item.get('title', 'unknown')}: {e}")
                results.append(
                    ExtractionResult(
                        summary=f"Extraction failed: {e}",
                    )
                )

        return results

File: weekly-intel-web/src/lib/services/processing/extractor.ts
import Anthropic from "@anthropic-ai/sdk";
import { config } from "@/lib/config";
import {
  EXTRACTION_SYSTEM_PROMPT,
  buildExtractionUserPrompt,
} from "@/lib/prompts";

export interface ExtractionResult {
  summary: string;
  keyInformation: string[];
  themes: string[];
  hotTakes: { take: string; context: string }[];
  entities: {
    people?: string[];
    companies?: string[];
    technologies?: string[];
  };
}

const EXTRACTION_TOOLS: Anthropic.Tool[] = [
  {
    name: "extract_content",
    description: "Extract structured information from content",
    input_schema: {
      type: "object" as const,
      properties: {
        summary: {
          type: "string",
          description: "A concise 2-3 sentence summary of the content",
        },
        key_information: {
          type: "array",
          items: { type: "string" },
          description:
            "List of key facts, announcements, or data points",
        },
        themes: {
          type: "array",
          items: { type: "string" },
          description: "Main themes or topics discussed",
        },
        hot_takes: {
          type: "array",
          items: {
            type: "object",
            properties: {
              take: {
                type: "string",
                description: "The contrarian or notable opinion",
              },
              context: {
                type: "string",
                description: "Brief context about why this is notable",
              },
            },
            required: ["take", "context"],
          },
          description: "Contrarian views or notable opinions",
        },
        entities: {
          type: "object",
          properties: {
            people: {
              type: "array",
              items: { type: "string" },
              description: "People mentioned",
            },
            companies: {
              type: "array",
              items: { type: "string" },
              description: "Companies or organizations mentioned",
            },
            technologies: {
              type: "array",
              items: { type: "string" },
              description:
                "Technologies, products, or concepts mentioned",
            },
          },
          description: "Named entities extracted from content",
        },
      },
      required: [
        "summary",
        "key_information",
        "themes",
        "entities",
      ],
    },
  },
];

let anthropicClient: Anthropic | null = null;

function getClient(): Anthropic {
  if (!anthropicClient) {
    anthropicClient = new Anthropic({ apiKey: config.anthropicApiKey });
  }
  return anthropicClient;
}

export async function extractContent(
  content: string,
  title?: string,
  author?: string,
  sourceType?: string
): Promise<ExtractionResult> {
  const contextParts: string[] = [];
  if (title) contextParts.push(`Title: ${title}`);
  if (author) contextParts.push(`Author: ${author}`);
  if (sourceType) contextParts.push(`Source type: ${sourceType}`);
  const context = contextParts.join("\n");

  // Truncate content if too long
  const maxLength = 50000;
  const truncatedContent =
    content.length > maxLength
      ? content.slice(0, maxLength) + "\n[Content truncated...]"
      : content;

  const userMessage = buildExtractionUserPrompt(truncatedContent, context);
  const client = getClient();

  const response = await client.messages.create({
    model: config.claudeModel,
    max_tokens: 4096,
    system: EXTRACTION_SYSTEM_PROMPT,
    tools: EXTRACTION_TOOLS,
    tool_choice: { type: "tool" as const, name: "extract_content" },
    messages: [{ role: "user", content: userMessage }],
  });

  for (const block of response.content) {
    if (block.type === "tool_use" && block.name === "extract_content") {
      const data = block.input as Record<string, unknown>;
      return {
        summary: (data.summary as string) ?? "",
        keyInformation: (data.key_information as string[]) ?? [],
        themes: (data.themes as string[]) ?? [],
        hotTakes:
          (data.hot_takes as { take: string; context: string }[]) ?? [],
        entities:
          (data.entities as ExtractionResult["entities"]) ?? {},
      };
    }
  }

  // Fallback
  const textContent = response.content
    .filter((b): b is Anthropic.TextBlock => b.type === "text")
    .map((b) => b.text)
    .join("");

  return {
    summary: textContent.slice(0, 500) || "Failed to extract summary",
    keyInformation: [],
    themes: [],
    hotTakes: [],
    entities: {},
  };
}
